### DISCUSSION TITLE
Choosing stopping criteria before starting the review process

### DISCUSSION THREAD

**Discussion by juliandwain** [asreview/asreview#1115] [Ideas]
Hi,

Thanks for this amazing project!
I have two questions, the first one is more a general question, and the second one is an idea I got while using ASReview.
I wanted to share this idea and get your feedback on it!

1. Is there a prospective date when v1 will be released? (is there a timeline or something similar?) (in my case, I am looking forward to the idea proposed in #1089.

Now onto the idea.
In #557, stopping criteria are categorized by estimating the number of relevant papers or using heuristics. 
(VAN HAASTRECHT ET AL. 2021) propose a simple estimate how to guess the number of relevant papers:

$$ R \approx N \cdot  {r \over r+i}$$

Herein, $R$ is the estimated number of relevant papers, $N$ is the total number of records in the dataset, $r$ is a pre-defined number of papers to mark as relevant, and $i$ denotes the number of papers marked as irrelevant at the current stage.
The screening can be stopped if a pre-defined percentage (e.g. 95 %) of the estimated number of relevant papers $R$ has been marked relevant.
There certainly are more methods to guess the number of relevant papers.

However, my suggestions would be that the user can decide what criteria to follow, i.e., estimating the number of relevant papers or using heuristics (a combination of both could also be beneficial) before starting the review process.
ASReview could then provide an overview of how close the user is to the stopping criteria, e.g.,

- heuristics (I think this is already sufficiently implemented, but when choosing a time-based method, then maybe a timer can be useful)
- estimating the number of relevant papers: herein, considering the formula from above, what might be useful is showing how much the number of papers to be marked as relevant *r* and the estimated number of relevant papers *R* are different from another - what is then shown to the user is how close one is to the defined percentage from above

I am happy to hear your thoughts on this and I hope that this topic is/was not already discussed elsewhere!

Best,
J

### References

VAN HAASTRECHT ET AL. 2021
van Haastrecht, M.; Sarhan, I.; Yigit Ozkan, B.; Brinkhuis, M.; Spruit, M.: SYM-BALS: A Systematic Review Methodology Blending Active Learning and Snow-balling. Frontiers in Research Metrics and Analytics 6 (2021), S. 1‑14.

**Comment by Rensvandeschoot** [comment]
> Is there a prospective date when v1 will be released? (is there a timeline or something similar?)

Yes! we aim at June 7-10... see the [milestones overview](https://github.com/asreview/asreview/milestone/3) for topics we are still working on.

**Comment by Rensvandeschoot** [comment]
@MaxvanHaastrecht: How would you propose to define **r**??

**Reply by MaxvanHaastrecht** [reply]
I fully agree with @juliandwain that it should always be up to the user to choose their stopping criterion. I think the discussion in #557 gives nice insight into how (for example) the time constraints on the researcher and the expected percentage of relevant documents will determine what type of stopping criterion is most appropriate. So ASReview should never make it mandatory to choose a particular stopping criterion.

That being said, I like the idea of @juliandwain to have an option in ASReview to display estimates related to your chosen stopping criterion strategy somewhere. When using static estimates defined a priori, it would be fairly easy to display this in some way. However, given the context of ASReview, it might be more fitting to incorporate dynamic estimates which update during the review. This does not imply that you can choose during the review what the strategy for obtaining this estimate would be (that goes against the spirit of a systematic review!). You would have to choose a dynamic strategy beforehand, and stick to it.

Note that not all (dynamic) stopping criterion strategies involve an estimate of the total number of relevant papers R. The 'knee' method of Cormack & Grossman (2016, https://doi.org/10.1145/2911451.2911510) is an example. So you would additionally have to see which estimates you would display exactly for which stopping criteria strategies. For discussions of alternative dynamic estimation strategies, see for example Li & Kanoulas (2020): https://doi.org/10.1145/3411755, and Callaghan & Müller-Hansen (2020): https://doi.org/10.1186/s13643-020-01521-4. Li & Kanoulas (2020) include a section specifically focused on estimating R, Section 5.2.

In short, I think it is a good idea to investigate how this facet could be incorporated in ASReview. I think the choice of what to incorporate and how is not trivial. It warrants ASReview-specific research that looks at which strategies (static vs dynamic) offer accurate estimates in which situations, but also which strategies are understandable and interpretable by users. There should be at least one 'simple' option in my opinion, which acts as a robust default.

**Reply by Rensvandeschoot** [reply]
In v1.0 we will offer more information you could use for the stopping rule. The information used for both 
graphs can be exported as CSV. 

![image](https://user-images.githubusercontent.com/36502709/170041888-da161925-7dbb-4ce3-91eb-d1bed3c885ee.png)

**Reply by juliandwain** [reply]
Thank you very much for your feedback!
I really like the idea of including a 'robust' default.

Also, I think the Analytics page in v1.0 looks really good!
Keep up the good work 👍

**Reply by Rensvandeschoot** [reply]
Would it be possible to apply the 'knee' method to the data from the lowest graph? Because if so, you can easily build a plug-in that reads this file and produces the criterion. Example data (fully labelled dataset):
[ASReviewLABprogressRecall (2).csv](https://github.com/asreview/asreview/files/8763607/ASReviewLABprogressRecall.2.csv)

**Reply by MaxvanHaastrecht** [reply]
@Rensvandeschoot I tried implementing the algorithm using the data, and I think the answer is yes. The result I get with the knee algorithm is to stop after reviewing 432 papers with your dataset. This would achieve a recall of 84% (although you could normally not calculate this value, since you do not know the total number of relevant papers).

See the attached PDF for the details. If you want me to share the notebook, let me know. I can't give any guarantees as to the veracity of this implementation, as I'm not completely sure whether I interpreted the Cormack & Grossman approach correctly.
[Cormack Grossman Knee Algorithm.pdf](https://github.com/asreview/asreview/files/8765859/Cormack.Grossman.Knee.Algorithm.pdf)

**Reply by Rensvandeschoot** [reply]
super nice!!! We did not implement this feature for creating stopping rules, but it might be a nice side-effect!!

**Reply by MaxvanHaastrecht** [reply]
Yeah I think there are definitely possibilities here. I shared the (updated) code here: https://github.com/MaxvanHaastrecht/ASReview-Knee-Method/blob/main/Cormack%20Grossman%20Knee%20Algorithm.ipynb. Also for @juliandwain, feel free to use it and/or improve on it.

**Reply by juliandwain** [reply]
@MaxvanHaastrecht thank you very much! I will definitely have a look at it

**Reply by j0sien** [reply]
Thanks @MaxvanHaastrecht , it looks great. I just do not understand yet how many labeled records one would need in order to work with the Knee Algorithm?

**Reply by MaxvanHaastrecht** [reply]
Hey @j0sien, the Knee algorithm has a parameter '_k_', the minimum number of relevant papers you want to find, which you have to choose before starting. The main reason for using this parameter is that since you are comparing two slopes (below the knee and above the knee), you want to make sure that the 'below the knee' part is not too short to avoid any extreme scenarios where you would stop very early on.

To answer your question, for the knee algorithm it doesn't matter much what your starting 'random' set is. You could screen randomly using ASReview until you have reached this number _k_ of relevant papers. You could also start once you have found (e.g.) at least one relevant paper, so that at least you know ASReview can do its job. I am assuming _k_>1, Cormack & Grossman suggest _k_=10, but this depends a bit on your dataset. The knee algorithm is then applied dynamically while you screen, meaning once every few iterations (or whenever you think the slope is starting to flatten off) you would need to check whether and when you have reached the 'knee'.

Hope this helps!

**Reply by tychovt99** [reply]
Dear @MaxvanHaastrecht , thank you for the ASReview Knee Method code. I'm trying to use it to find a stopping criterion for my current systematic review, but I unfortunately do not have very much experience in Python. 

I was wondering how you downloaded the .csv file you used in this code, as the columns are very different from my dataset that I download from ASReview. For example, I do not have "category" and "Relevant by ASReview Lab". My dataset includes authors, title, doi, abstract, keywords, included, and asreview_ranking. From what I understand, 'included' indicates whether I have checked a paper as relevant or irrelevant within ASReview. I'm not sure how I can translate my dataset's columns into your code, do you think you could give me some advice on this? Any help would be appreciated.

Cheers,
Tycho

**Reply by MaxvanHaastrecht** [reply]
Hey @tychovt99, the Python script was made to work based on the CSV @Rensvandeschoot provided in this comment: https://github.com/asreview/asreview/discussions/1115#discussioncomment-2812635. If you go to the Analytics tab of the project you're working on and download the CSV of the Progress graph (see attached image), you should get a CSV with a "category" and "Relevant records" column. These correspond to the "category" and "Relevant by ASReview Lab" columns from the original example.

To be clear, calculating the knee stopping criterion is something you do during the review process. In practice, you might want to stop every 20 records or so and see if your stopping criterion has been satisfied. This avoids having to do the calculation after every decision you make.

![download csv](https://user-images.githubusercontent.com/65038675/220938597-283599f0-f654-4a03-b9e7-8754f8f6d607.jpg)

**Reply by tychovt99** [reply]
Thank you for the quick and informative reply! I managed to find the CSV file this way and it seems to work well now. Also thank you for the suggestion on stopping, my supervisor and I were considering stopping after batches of 5% of the total records.

**Reply by Rensvandeschoot** [reply]
@tychovt99 see also the proposed stopping heuristic by @j0sien in this thread: https://github.com/asreview/asreview/discussions/1300
